### PR TITLE
Add configurable filters to mirrors

### PR DESCRIPTION
**Pull request by psakievich**
Add the ability to filter what specs get pushed to a mirror. This allows for a more controlled autopush for binary mirrors. 

Currently filtering can only be called on `spack mirror create` and `spack mirror create` will only create a source mirror and NOT a binary mirror. Refactoring all of that I considered out of scope for this effort since that would be a refactor of most of the module and need a rethink on the user API.

Since source mirror creation also does population the filtering only gets preserved during the lifetime of the method.

For binary mirrors we need the filters to last during a `spack install` for `--autopush` so the filter specs need to be preserved. The implementation in this PR is to recorded the filter specs for binary mirrors in the `mirrors.yaml` because I am considering them a property of the mirror.

Requested to not merge until after #48713

### CODE REVIEW

File: lib/spack/spack/mirrors/filter.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    def __call__(self, specs: List[spack.spec.Spec]):
+        """
+        Determine the intersection of include/exclude filters
+        Tie goes to keeping
+
+        skip  | keep  | outcome
+        ------------------------
+        False | False | Keep
+        True  | True  | Keep
```
Comment (kwryankrattiger) [inline lib/spack/spack/mirrors/filter.py:23]: I think this is okay, but I think calling the second field include is kind of strange since things that don't match the include are also included. Adding a `@:` match for include disables the exclude, which makes the meaning of exclude kind of wonky.

Really this is a exclude, and override/exempt from exclude, which is a little different than include. Maybe for each exclusion there could be an optional exemptions field?

```yaml
exclude:
    - match: dev_path=*
      exempt:
          - zlib
          - ~shared
    - intel-oneapi-runtime
```

Comment (haampie) [reply lib/spack/spack/mirrors/filter.py:23]: just be consistent with views/modules etc. https://spack.readthedocs.io/en/latest/environments.html#advanced-view-configuration

Comment (kwryankrattiger) [reply lib/spack/spack/mirrors/filter.py:23]: The main issue with that is to get detailed exclusions in a dev environment, where the majority of dev specs should be excluded, is you need to use `<package_name> dev_path=*` to say you want to push one spec to a shared developer build cache. That is potentially a lot of config if there are 10s of dev specs in the environment and leads to more potential for error.

Granted, a tool like `spack-manager` could maintain a reference block in `mirrors.yaml` that managed the full list of develop specs maybe that could work...

```yaml
.spack-manager:
  develop-specs: &sm_mirror_exclusions
    - zlib dev_path=*
    - ncurses dev_path=*

mirrors:
  my_mirror:
     select: "@:"
     exclude:
     - *sm_mirror_exclusions
     - intel-oneapi-runtime
```

Comment (haampie) [reply lib/spack/spack/mirrors/filter.py:23]: same problem for modules and views...

Comment (haampie) [reply lib/spack/spack/mirrors/filter.py:23]: https://spack.readthedocs.io/en/latest/module_file_support.html#exclude-or-include-specific-module-files looks like views and modules are already inconsistent and the module way is what you want?

Comment (psakievich) [reply lib/spack/spack/mirrors/filter.py:23]: @kwryankrattiger I added the `--exclude-from-file` to `spack mirror add` so complex exclusion configurations can be supported. Oh but I suppose what you want is `--include-from-file`.  I can add that.

@haampie  I __think__ what I have matches modules. @kwryankrattiger I agree exemption is likely a better name. I can make that change here, but it is adding another bifurcation of syntax.

Comment (psakievich) [reply lib/spack/spack/mirrors/filter.py:23]: Overall I would prefer to not add a more advanced syntax here. I think between spec matrices, views and modules we should refine and consolidate, then be consistent here. Straight lists are a neutral party between all these cases.

Comment (kwryankrattiger) [reply lib/spack/spack/mirrors/filter.py:23]: I think following existing configs is a better course than inventing something new. The file support is fine to handle these more complex cases, the perennial problem then becomes more tooling to be able to generate those files.

**Comment by psakievich** [comment]
Converting to draft until #48713 merges

**Comment by psakievich** [comment]
@haampie @kwryankrattiger bump

**Comment by psakievich** [comment]
Mega squash to satisfy DCO

File: lib/spack/spack/cmd/mirror.py [hunk 2]
```diff
@@ -428,6 +491,8 @@ def _configure_mirror(args):
         changes["source"] = "source" in args.type
 
     changed = entry.update(changes, direction)
+    if hasattr(args, "include_file"):
```
Comment (kwryankrattiger) [inline lib/spack/spack/cmd/mirror.py:460]: I don't see why this check is needed.

File: lib/spack/docs/mirrors.rst [hunk 3]
```diff
@@ -120,6 +120,16 @@ what it looks like:
 Once this is done, you can tar up the ``spack-mirror-2014-06-24`` directory and
 copy it over to the machine you want it hosted on.
 
+Customization of the mirror contents can be done by selectively excluding
+specs using the ``--exclude-file`` or ``--exclude-specs`` flags with
+``spack mirror create``.  Note that these only apply to source mirrors.
```
Comment (kwryankrattiger) [inline lib/spack/docs/mirrors.rst]:
```suggestion
``spack mirror [add|set]``.
```

These options are added for `spack mirror add` and `spack mirror set` but not `spack mirror create`.

File: lib/spack/spack/cmd/mirror.py [hunk 4]
```diff
@@ -299,6 +335,30 @@ def _default_variable(id_):
         return None
 
 
+def _manage_filters(args, mirror) -> bool:
```
Comment (kwryankrattiger) [inline lib/spack/spack/cmd/mirror.py]: Nit on naming, I think maybe `_collect_mirror_filters` or `_update_mirror_filters` may be better.

And pass mirror first (since that is what is being operated on) and then args.

File: lib/spack/spack/mirrors/mirror.py [hunk 5]
```diff
@@ -97,6 +97,11 @@ def display(self, max_len=0):
         binary = "b" if self.binary else " "
         print(f"{self.name: <{max_len}} [{source}{binary}] {url}")
 
+    def _process_spec_filters(self, key: str) -> List[str]:
```
Comment (kwryankrattiger) [inline lib/spack/spack/mirrors/mirror.py]: This should follow the naming convention we use for the other fields 

```suggestion
    def _get_spec_filters(self, key: str) -> List[str]:
```

File: lib/spack/spack/schema/mirrors.py [hunk 6]
```diff
@@ -77,6 +77,8 @@
         "fetch": fetch_and_push,
         "push": fetch_and_push,
         "autopush": {"type": "boolean"},
+        "exclude": {"type": "array", "items": {"type": "string"}},
```
Comment (kwryankrattiger) [inline lib/spack/spack/schema/mirrors.py]: should these also support the file or list?

```
exclude: spec_exclusion_file.txt
```
or
```
exclude:
- intel-oneapi-mkl
- ...
```

**Review by kwryankrattiger — COMMENTED** [review]
Took another pass on the implementation, still looking at the tests.

File: lib/spack/spack/cmd/mirror.py (hunk 2 above)
Comment (psakievich) [reply lib/spack/spack/cmd/mirror.py:460]: The method is called to handle parsing for `set-url` and `set`. It is only valid in `set`. It is ugly but I didn't want to refactor this whole file as part of the PR since it was pending on #48713.

File: lib/spack/docs/mirrors.rst (hunk 3 above)
Comment (psakievich) [reply lib/spack/docs/mirrors.rst]: `spack mirror create` populates source files when it runs so the two flags in the docs will only apply to source files. I suppose we could change that to these flags apply uniformly to source and binary, but when you run `spack mirror create` it will populate a source mirror and not a binary mirror.

I was attempting to document our (IMHO poor) API.

File: lib/spack/spack/schema/mirrors.py (hunk 6 above)
Comment (psakievich) [reply lib/spack/spack/schema/mirrors.py]: I made it so the file contents get added to the list in the yaml. I can change that though if we want to keep the file as the source of truth.

**Comment by psakievich** [comment]
@kwryankrattiger  I updated the description with some additional context on the feature's implementation and scope. I would argue our mirror command is confusing and bifurcates with `buildcache` too much. But refactoring all of that for this feature seems like massive scope creep to me.

File: lib/spack/spack/schema/mirrors.py (hunk 6 above)
Comment (psakievich) [reply lib/spack/spack/schema/mirrors.py]: I did it this way to not have to support a merge for a file and yaml, but I suppose the mirror definition would be one or the other this way so it would work.

File: lib/spack/docs/mirrors.rst (hunk 3 above)
Comment (kwryankrattiger) [reply lib/spack/docs/mirrors.rst]: My point here is you are adding these arguments to add/set but the doc update here doesn't reflect that.

File: lib/spack/spack/cmd/mirror.py (hunk 2 above)
Comment (kwryankrattiger) [reply lib/spack/spack/cmd/mirror.py:460]: ah I see now, the diff in github made it look liek this was in the `mirror_add`  function at first glance.

File: lib/spack/spack/schema/mirrors.py (hunk 6 above)
Comment (kwryankrattiger) [reply lib/spack/spack/schema/mirrors.py]: yeah I think it would be nice, this list can get quite long so not having it in the YAML is an option I am looking for.

File: lib/spack/docs/mirrors.rst (hunk 3 above)
Comment (psakievich) [reply lib/spack/docs/mirrors.rst]: I'll update the docs to see if I can explain it all better. Good catch. I'm surprised I didn't add the new flags with their commands.

File: lib/spack/spack/schema/mirrors.py (hunk 6 above)
Comment (psakievich) [reply lib/spack/spack/schema/mirrors.py]: Hmm I'm wondering if I should make the list a spec or file? Then you could have multiple files and/or a combination rather than list of specs or single file.  Thinking about parsing this

```
spack mirror set foo --exclude-specs a b c --exclude-file blah.txt
```

Right now this is valid. I suppose I can make them mutually exclusive, but I don't see why they have to be.

Comment (psakievich) [reply lib/spack/spack/schema/mirrors.py]: Eh NVM. I  updated to preserve the file and added mutual exclusivity. I think simple is best for now. It shouldn't be too hard to add additional flexibility, but I also don't think we need to over engineer this right now.

File: lib/spack/spack/cmd/buildcache.py [hunk 7]
```diff
@@ -460,13 +460,19 @@ def push_fn(args):
             tty.info("All specs are already in the buildcache. Use --force to overwrite them.")
         else:
             tty.info(
-                "The following {} specs were skipped as they already exist in the buildcache:\n"
-                "    {}\n"
-                "    Use --force to overwrite them.".format(
-                    len(skipped), ", ".join(elide_list([_format_spec(s) for s in skipped], 5))
-                )
+                f"The following {len(skipped)} specs were skipped as they already exist in the"
```
Comment (becker33) [inline lib/spack/spack/cmd/buildcache.py:463]:
```suggestion
                f"The following {llnl.util.string.plural(len(skipped), "spec")} were skipped as they already exist in the"
```
This will properly display "spec" if it's only one and "specs" for multiple.

File: lib/spack/spack/cmd/buildcache.py [hunk 8]
```diff
@@ -460,13 +460,19 @@ def push_fn(args):
...
+                " buildcache:\n"
+                f"    {', '.join(elide_list([_format_spec(s) for s in skipped], 5))}\n"
+                "    Use --force to overwrite them."
             )
 
+    if excluded:
+        tty.info(
+            f"The following {len(excluded)} specs were ommitted due from pushing to the"
```
Comment (becker33) [inline lib/spack/spack/cmd/buildcache.py:471]: Same here for plural

File: lib/spack/spack/mirrors/filter.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        False | True  | Keep
+        True  | False | Skip
+        """
+        filter = []
+        filtrate = []
```
Comment (becker33) [inline lib/spack/spack/mirrors/filter.py:28]: This may be easier to read with `included` and `excluded` instead of `filter` and `filtrate`, respectively.

File: lib/spack/docs/mirrors.rst [hunk 10]
```diff
@@ -137,6 +137,19 @@ command:
 Will create a mirror for libelf versions greater than or equal to
 0.8.12 and boost versions greater than or equal to 1.44.
 
+Further customization of the mirror contents can be done by selectively excluding
+specs using the ``--exclude-file`` or ``--exclude-specs`` flags with
+``spack mirror create``.
+
+Note that these only apply to source mirrors when using ``spack mirror create``. 
+This is because ``spack mirror create`` only generates source mirrors.
```
Comment (becker33) [inline lib/spack/docs/mirrors.rst:145]: I'm not sure I understand what point this is making that is not made in the paragraphs on either side.

File: lib/spack/docs/mirrors.rst [hunk 11]
```diff
@@ -137,6 +137,19 @@ command:
...
+
+The ``--exclude-file`` and ``--exclude-specs`` flags may also be used with ``spack mirror [create | add]``. These will populate an ``exclude`` section in the mirrors configuration.
```
Comment (becker33) [inline lib/spack/docs/mirrors.rst:147]: I think this sentence is meant to only discuss the `spack mirror add` command, since the `create` command is already covered and doesn't save the options to config. Am I missing something?

File: lib/spack/spack/cmd/mirror.py [hunk 12]
```diff
@@ -134,6 +134,7 @@ def setup_parser(subparser: argparse.ArgumentParser) -> None:
         default=None,
         dest="signed",
     )
+    arguments.add_filter_args(add_parser)
```
Comment (becker33) [inline lib/spack/spack/cmd/mirror.py:137]: I think there should be analogous arguments on the `spack buildcache push` command, for the case that you want to restrict an individual command rather than restricting the entire mirror (analogous to the new `spack mirror create` arguments).

File: lib/spack/spack/cmd/mirror.py [hunk 13]
```diff
@@ -299,6 +301,30 @@ def _default_variable(id_):
         return None
 
 
+def _collect_mirror_filters(mirror, args) -> bool:
```
Comment (becker33) [inline lib/spack/spack/cmd/mirror.py:304]: Can you rename this to `_parse_mirror_filters` or something similar? `_collect` to me implies that we're returning a collection, which isn't true in this case.

File: lib/spack/spack/cmd/mirror.py [hunk 14]
```diff
@@ -470,7 +501,10 @@ def specs_from_text_file(filename, concretize=False):
     with open(filename, "r", encoding="utf-8") as f:
         specs_in_file = f.readlines()
         specs_in_file = [s.strip() for s in specs_in_file]
-    return spack.cmd.parse_specs(" ".join(specs_in_file), concretize=concretize)
+    if concretize:
+        return spack.cmd.parse_specs(" ".join(specs_in_file), concretize=True)
+    else:
+        return spack.cmd.parse_specs(specs_in_file, concretize=False)
```
Comment (becker33) [inline lib/spack/spack/cmd/mirror.py:507]: Why the different argument handling for these two cases?

File: lib/spack/spack/cmd/common/arguments.py [hunk 15]
```diff
@@ -632,6 +632,29 @@ def add_concretizer_args(subparser):
     )
 
 
+def add_filter_args(subparser):
```
Comment (becker33) [inline lib/spack/spack/cmd/common/arguments.py:635]: It would be nice if these didn't need to be exclusive groups (so someone could e.g. exclude a file for a common set of things they exclude, but also exclude an additional spec they need to exclude this time). It looks like that would require minor changes to the config format (an `exclude` block, with separate keys for `specs` and `files`, instead of a single entry under `exclude`) but otherwise seems feasible. Do you think that is workable?

**Review by becker33 — CHANGES_REQUESTED** [review]
Did not review the tests in detail -- generally, the tests look reasonable but they would need to change substantially for the other changes I've requested to cover new corner cases, so I left them out of this review.

**Comment by thomas-bouvier** [comment]
What is the status of this PR @psakievich ?

**Comment by psakievich** [comment]
Thanks for the bump @thomas-bouvier. Status is I saw @becker33's requests, didn't have capactiy to grow the feature, then ... I forgot about it.

I can revisit it if you have a need/desire for it.

File: lib/spack/spack/cmd/mirror.py (hunk 14 above)
Comment (psakievich) [reply lib/spack/spack/cmd/mirror.py:507]: IIRC the parser when concretizing input specs needs you to rejoin the specs for the solve.

File: lib/spack/spack/mirrors/filter.py (hunk 9 above)
Comment (psakievich) [reply lib/spack/spack/mirrors/filter.py:28]: Point taken. I didn't use those terms to avoid overloading since the filter can be used for inclusion/exclusion. I can change the names if the context is sufficiently clear.

File: lib/spack/docs/mirrors.rst (hunk 10 above)
Comment (psakievich) [reply lib/spack/docs/mirrors.rst:145]: True, I might be over-emphasizing the point. Personally, I find the lack of symmetry between create/add plus the interplay of the buildcache command to be very confusing.

File: lib/spack/docs/mirrors.rst (hunk 11 above)
Comment (psakievich) [reply lib/spack/docs/mirrors.rst:147]: That's a good point. I tried to make the flags mirror `create` for common look and feel, but the side-effects of modifying the config is different.

**Comment by thomas-bouvier** [comment]
Spack is becoming increasingly popular in French data centers, and we would like to have more flexibility around binary caches. This feature would be very useful :)

File: lib/spack/spack/cmd/common/arguments.py (hunk 15 above)
Comment (psakievich) [reply lib/spack/spack/cmd/common/arguments.py:635]: Sigh. This is a fair point. I just didn't want to reimplement this extension. The logic for interleaving the specs from different sources plus include and exclude was more complicated to implement than it seems.

Every other request in this review is pretty straightforward. This one I think will take more work. I can take another pass at this. Maybe using an LLM it will go super fast since this is 90% done already. If not can we define an intermediate state and expand this later? My main concern is it is changing config which we want to not change too often.

Comment (psakievich) [reply lib/spack/spack/cmd/common/arguments.py:635]: @thomas-bouvier want to weigh in on this? Is the change requested a significant impact to your use-cases?

Comment (thomas-bouvier) [reply lib/spack/spack/cmd/common/arguments.py:635]: I don't have a strong opinion on this. Maybe this feature could be a follow-up PR.

**Comment by thomas-bouvier** [comment]
I am not entirely sure of the following scenario, but I think it happened to me some time ago: dependencies of an environment marked as "develop" (using `spack develop`) ended up in a binary cache, which is certainly not intended.

@psakievich would you confirm this scenario is possible? If it can happen, maybe an argument should be added to exclude "develop" dependencies by default? Or maybe this should be addressed in a follow-up PR?

**Comment by psakievich** [comment]
@thomas-bouvier yes that is possible. We do exclude develop specs from reuse so you probably can't pull them back unless it is an exact hash match but you also don't want to push them in the first place. Also even if they are an exact hash match you almost certainly wouldn't want the binary from a build cache since you're developing it.

**Comment by thomas-bouvier** [comment]
Would it make sense to filter out develop dependencies by default? This would avoid polluting binary caches

**Comment by psakievich** [comment]
> Would it make sense to filter out develop dependencies by default? This would avoid polluting binary caches

There was another effort that was being explored to make dev specs in a binary cache a feature. This was to allow developers to reuse each other's binaries in a fast, iterative manner. I think it is mostly stalled, but that is why I made a more general feature.
